File: 20220921_Wednesday/20220921.go
// Leetcode: 985. Sum of Even Numbers After Queries.
// https://leetcode.com/problems/sum-of-even-numbers-after-queries/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = = 
// Runtime: 179 ms, faster than 50.00% of Go online submissions for Sum of Even Numbers After Queries.
// Memory Usage: 7.6 MB, less than 64.29% of Go online submissions for Sum of Even Numbers After Queries
func sumEvenAfterQueries(nums []int, queries [][]int) []int {
  answer := make([]int, len(queries))
  total := 0
  // Iterate over nums and get sum of even values.
  for _, val := range nums {
    if 0 == val%2 {
      total += val
    }
  }
  // Iterate over queries and calculate new value of the total.
  var  value, index int
  for query_i, pair := range queries {
    value, index = pair[0], pair[1]
    if 0 == nums[index]%2 {
      total -= nums[index]
    }
    nums[index] += value
    if 0 == nums[index]%2 {
      total += nums[index]
    }
    answer[query_i] = total
  }
  return answer
}

File: 20220921_Wednesday/20220921.rb
# Leetcode: 985. Sum of Even Numbers After Queries.
# https://leetcode.com/problems/sum-of-even-numbers-after-queries/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 341 ms, faster than 100.00% of Ruby online submissions for Sum of Even Numbers After Queries.
# Memory Usage: 215.1 MB, less than 100.00% of Ruby online submissions for Sum of Even Numbers After Queries.
# @param {Integer[]} nums
# @param {Integer[][]} queries
# @return {Integer[]}
def sum_even_after_queries(nums, queries)
  @d = false
  @answer = []
  @n, @q = nums, queries
  @sum = nums.map{|x| x.even? ? x : 0 }
  puts "Sum and total." if @d
  p @sum if @d
  total = @sum.sum
  p total if @d
  answer = []
  @q.each_with_index do |(val,i),ii| # Fixed Error: Was (i,v), need (v,i).
    total -= @sum[i]
    puts "New total #{total} after remove #{@sum[i]}." if @d
    @n[i] += val
    @sum[i] = @n[i].even? ? @n[i] : 0
    p @sum if @d
    total += @sum[i]
    puts "Modifed total is #{total} after add #{@sum[i]}." if @d
    answer.push(total)
  end
  answer
end

